### DISCUSSION TITLE
Custom trained POS tagger pipeline tags pos as empty string.

### DISCUSSION THREAD

**Discussion by Rolands-Laucis** [explosion/spaCy#7793] [Help: Coding & Implementations] [answered]
## Which page or section is this issue related to?
https://spacy.io/usage/spacy-101#training

I trained a custom pipeline with tok2vec and tagger components. I want to do POS tagging for the Latvian language using word embeddings, that i've generated prior. I used the tool on [the usage site (https://spacy.io/usage/training) to generate the base_config file. Then went through the steps on CLI.
Filling in the defaults for config:
`python -m spacy init fill-config base_config.cfg config.cfg`
Converting my .conllu training files to .spacy:
`python -m spacy convert lv_lvtb-ud-test.conllu .\ --converter conllu`
Converting my embeddings into .spacy:
`python -m spacy init vectors lv ../../Models/SSG_model/ssg_5_200_sg.txt ../../Models/Spacy_tagger/vectors --name lv-ssg-5-200`
Then in that dir in the config added that output path under [Initialize] vectors = ...
Then ran the training like so:
`python -m spacy train config.cfg --output ../../Models/Spacy_tagger/ssg_5_200_pos --paths.train ./lv_lvtb-ud-train.spacy --paths.dev ./lv_lvtb-ud-dev.spacy`
Then in python run this code:
`import spacy

nlp = spacy.load("../../Models/Spacy_tagger/ssg_5_200_pos/model-last/")
doc = nlp("Es pārbaudu kā strādā šis kods.")

print(nlp.pipeline)

for token in doc:
    #print(token.text, token.pos)
    print(token.text, token.pos_)`

That outputs:
[('tok2vec', <spacy.pipeline.tok2vec.Tok2Vec object at 0x0000029654EA17C0>), ('tagger', <spacy.pipeline.tagger.Tagger object at 0x0000029654F8BE00>)]
Es
pārbaudu
kā
strādā
šis
kods 
.

Clearly the POS property of the token is "" and doing token.pos prints all of them 0.

What am i doing wrong? There were no errors in the training or other steps. I have read and reread the docs and dont understand. During training the accuraccy of tagger was 77% at the end.

**Comment by adrianeboyd** [comment, marked as answer]
Hi, spacy supports two different types of part-of-speech tags: fine-grained POS in `token.tag` which can be any string and coarse-grained POS in `token.pos` which is restricted to only the UD POS tag values.

The `tagger` component assigns `token.tag`, which is converted from the XPOS column (fine-grained tag) in the UD corpus. If you inspect `token.tag_` instead above, you'll see the fine-grained tags from the UD corpus.

If you want `token.pos` add a `morphologizer` component, which assigns `token.pos` and `token.morph` trained on the UPOS and FEATS CoNLL-U columns in the converted corpus.

**Reply by Rolands-Laucis** [reply]
Thank you for your reply, it is now a lot clearer to me. I indeed see the token.tag_ names printed. In the [Tagger API](https://spacy.io/api/tagger) it is stated that 

> Pipeline component for part-of-speech tagging

And so i thought thats all i needed for my case. 

A follow-up question though. The [morphology usage docs](https://spacy.io/usage/linguistic-features#morphology) dont specify how to add this component in the config.cfg file. And the [morphologizer API](https://spacy.io/api/morphologizer) explains it can be set in code, which i wont do because i am training my model through CLI not through code, as well as can be set in the config file, but links to the config file options general overview. Would it be something like [morphologizer] under [components]? And if so, what are its defaults to be set? Is there a documentation page i was not able to find, that has this information?

Thank you for your time, this is really helping me with the final stages of my thesis.

**Reply by adrianeboyd** [reply]
Ah, it's missing from the quickstart, we'll get that updated.

You can create a config including any component by specifying it with the `-p` option with `spacy init config`:

```bash
spacy init config -p morphologizer config.cfg
```

If you want other components, provide them as a list:

```bash
spacy init config -p tagger,morphologizer,parser config.cfg
```

You can also have a look at the project for [training from UD corpora](https://github.com/explosion/projects/tree/v3/benchmarks/ud_benchmark):

```bash
spacy project clone benchmarks/ud_benchmark
```

It's specifically set up to run benchmarks for UD v2.5, which might not be exactly what you want to do, but you can see the commands to convert, generate a config, and train.
